pulp_ansible/app/serializers.py

```diff
@@ -361,130 +373,259 @@ class Meta:
+            if not all((f"expected_{x}" in data for x in fields)):
+                if "file" not in data:
+                    raise serializers.ValidationError(
+                        _(
+                            "expected_namespace, expected_name, and expected_version must be "
+                            "specified when using artifact or upload objects"
+                        )
+                    )
+                filename = data["file"].name
+                try:
+                    collection = parse_collection_filename(filename)
+                except ValueError:
+                    raise ValidationError(
+                        _("Failed to parse Collection file upload '{}'").format(filename)
+                    )
+                data["expected_namespace"] = collection.namespace
+                data["expected_name"] = collection.name
+                data["expected_version"] = collection.version
+
+            if CollectionVersion.objects.filter(**{f: data[f"expected_{f}"] for f in fields}):
+                raise serializers.ValidationError(
+                    _("Collection {}.{}-{} already exists").format(
+                        data["expected_namespace"], data["expected_name"], data["expected_version"]
+                    )
+                )
 
-        sha256 = data["file"].hashers["sha256"].hexdigest()
-        artifact = Artifact.objects.filter(sha256=sha256).first()
-        if artifact:
-            ValidationError(_("Artifact already exists"))
-        temp_file = PulpTemporaryFile.init_and_validate(data.pop("file"))
-        temp_file.save()
-        data["temp_file_pk"] = str(temp_file.pk)
+        return data
+
+    def deferred_validate(self, data):
+        """Import the CollectionVersion extracting the metadata from its artifact."""
+        # Call super to ensure that data contains artifact
+        data = super().deferred_validate(data)
+        artifact = data.get("artifact")
+        if (sha256 := data.pop("expected_sha256", None)) and sha256 != artifact.sha256:
+            raise serializers.ValidationError(
+                _("Expected sha256 did not match uploaded artifact's sha256")
+            )
+
+        collection_info = process_collection_artifact(
+            artifact=artifact,
+            namespace=data.pop("expected_namespace"),
+            name=data.pop("expected_name"),
+            version=data.pop("expected_version"),
+        )
+        # repository field clashes
+        collection_info["origin_repository"] = collection_info.pop("repository", None)
+        data.update(collection_info)
 
         return data
 
+    def create(self, validated_data):
+        """Final step in creating the CollectionVersion."""
+        tags = validated_data.pop("tags")
+        origin_repository = validated_data.pop("origin_repository")
+        # Create CollectionVersion from its metadata and adds to repository if specified
+        content = super().create(validated_data)
```

Comment:
Does it make sense to add the add-to-repository-in-a-subtask feature to pulpcore?

Reply:
I'm thinking about it. I already know how I am going to implement the feature in pulp_ansible, but putting it into pulpcore would make it much more simple.